sips/sip-021/sip-021-nakamoto.md

```diff
@@ -0,0 +1,1040 @@
+to include them.  In other words, Stackers _always_ produce a `TenureChange`
+transaction for each tenure; it just might not be included by the tenure's
+sortition-selected miner.
+
+### Stacker Turnover
+
+A miner tenure change happens every Bitcoin block.  In addition, there are
+Stacker tenure changes, which happen once every 2100 Bitcoin blocks (one reward
+cycle).  In a Stacker tenure change, a new set of Stackers are selected by the
+PoX anchor block (see SIP-007).
+
+Because Stacks no longer forks, the PoX anchor block is always known 100 Bitcoin
+blocks prior to the start of the next reward cycle:  it is simply the the start
+block of the 2000th tenure in the current reward cycle.
+
+The PoX anchor block identifies the next Stackers.  They have 100 Bitcoin blocks
+to prepare for signing Stacks blocks.  Within this amount of time, the new
+Stackers would complete a WSTS DKG for signing blocks (among other things, such
+as the sBTC peg wallet hand-off).  The PoX contract will require Stackers to
+register their block-signing keys when they stack or delegate-stack STX, so the
+entire network knows enough information to validate their WSTS Schnorr
+signatures on blocks.
+
+### PoX Failure
+
+In the event that PoX does not activate, the chain halts.  If there are no
+Stackers, then block production cannot happen.
+
+### Changes to PoX
+
+To support tenure changes, this proposal calls for a new PoX contract, `.pox-4`.
+The `.pox-4` contract would be altered over the current PoX contract (`.pox-3`)
+to meet the following requirements:
+
+* Stackers register a WSTS signing key when they call `stack-stx` and optionally
+  `delegate-stx`.
+    * In `delegate-stx`, Stackers may instead elect for the delegate to supply
+      the signing key.  In this case, the delegate carries out the Stacker's
+responsibility for signing blocks.
+* The stacking minimum is abolished.  If there is at least one stacker that is
+  able to claim one reward slot, then PoX activates.
+* The `.pox-4` contract will expose each reward cycle's full reward set,
+  including the signing keys for each stacker, via a public function.
+Internally, the Stacks node will call a _private_ function to load the reward
+set into the `.pox-4` data space after it identifies the PoX anchor block.  This
+is required for other dependent SIPs, such as the sBTC SIP.
+
+## Dynamic Runtime Budgets
+
+The time between cryptographic sortitions (and thus tenure changes) depends on
+the time between two consecutive burnchain blocks.  This can be highly variable,
+which complicates the task of sustaining a predictable transaction confirmation
+latency while also preventing a malicious miner from spamming the network with
+too many high-resource transactions.
+
+Today, each miner receives a tenure block budget, which places hard limits on
+how much CPU, RAM, and I/O their block can consume when evaluated.  In this
+proposal, each miner begins its tenure with a fixed budget, but Stackers may opt
```

Comment:
This would be equivalent to the proposed system.  Either way, Stackers need to be able to throttle the miner's compute usage -- either by explicitly granting the miner an extension, or by explicitly preventing the miner from consuming resources too quickly (such as by NACKing their block or individual transactions within them).  But in both cases, the system must impose a fixed rate of resource consumption.